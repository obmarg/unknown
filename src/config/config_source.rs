use std::{fmt, sync::Arc};

use camino::Utf8PathBuf;
use miette::MietteSpanContents;

#[derive(Clone)]
pub struct ConfigSource {
    inner: Arc<Inner>,
}

impl std::fmt::Debug for ConfigSource {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("ConfigSource")
            .field("filename", &self.inner.filename)
            .finish_non_exhaustive()
    }
}

struct Inner {
    filename: String,
    code: String,
}

impl ConfigSource {
    pub fn new(filename: impl AsRef<str>, code: impl Into<String>) -> Self {
        ConfigSource {
            inner: Arc::new(Inner {
                filename: filename.as_ref().to_string(),
                code: code.into(),
            }),
        }
    }

    // pub fn to_miette(&self) -> miette::NamedSource {
    //     miette::NamedSource::new(&self.inner.filename, self.inner.code.clone())
    // }
}

impl miette::SourceCode for ConfigSource {
    fn read_span<'a>(
        &'a self,
        span: &miette::SourceSpan,
        context_lines_before: usize,
        context_lines_after: usize,
    ) -> Result<Box<dyn miette::SpanContents<'a> + 'a>, miette::MietteError> {
        let contents =
            self.inner
                .code
                .read_span(span, context_lines_before, context_lines_after)?;

        Ok(Box::new(MietteSpanContents::new_named(
            self.inner.filename.clone(),
            contents.data(),
            *contents.span(),
            contents.line(),
            contents.column(),
            contents.line_count(),
        )))
    }
}
