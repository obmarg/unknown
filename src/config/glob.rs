use knuffel::{
    ast::Literal, decode::Kind, errors::DecodeError, span::Spanned, traits::ErrorSpan, DecodeScalar,
};
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct Glob(globset::Glob);

impl Glob {
    pub fn into_inner(self) -> globset::Glob {
        self.0
    }
}

impl<S> DecodeScalar<S> for Glob
where
    S: ErrorSpan,
{
    fn type_check(
        _type_name: &Option<knuffel::span::Spanned<knuffel::ast::TypeName, S>>,
        _ctx: &mut knuffel::decode::Context<S>,
    ) {
        // Not bothering with types for now...
    }

    fn raw_decode(
        value: &Spanned<Literal, S>,
        _ctx: &mut knuffel::decode::Context<S>,
    ) -> Result<Self, DecodeError<S>> {
        let Literal::String(s) = &**value else {
            let found =  match **value {
                Literal::Null => Kind::Null,
                Literal::Bool(_) => Kind::Bool,
                Literal::Int(_) => Kind::Int,
                Literal::Decimal(_) => Kind::Decimal,
                Literal::String(_) => panic!("this should be impossible")
            };
            return Err(DecodeError::ScalarKind {
                span: value.span().to_owned(),
                expected: Kind::String.into(),
                found
            });
        };

        let glob = globset::Glob::new(s.as_ref()).map_err(|error| DecodeError::Conversion {
            span: value.span().to_owned(),
            source: Box::new(error),
        })?;

        Ok(Glob(glob))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(knuffel::Decode, Debug)]
    pub struct TestStruct {
        #[knuffel(children(name = "path"), unwrap(argument))]
        pub paths: Vec<Glob>,
    }

    #[test]
    fn test_decoding_globs() {
        let result = knuffel::parse::<TestStruct>(
            "whatevs.txt",
            r#"
        path "hello/**"
        path "**"
        path "*.txt"
        path "a_file.txt"
        "#,
        )
        .unwrap();

        insta::assert_debug_snapshot!(result);
    }
}
