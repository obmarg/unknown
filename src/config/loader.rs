use std::path::{Path, PathBuf};

use super::{parse_project_file, workspace_from_str, ParsingError, ProjectFile, WorkspaceFile};

pub fn load_config_from_cwd() -> Result<(WorkspaceFile, Vec<ProjectFile>), miette::Report> {
    let workspace_path = find_workspace_file().ok_or(MissingWorkspaceFile)?;
    let workspace_file = workspace_from_str(
        &std::fs::read_to_string(&workspace_path).expect("couldn't read workspace file"),
    )
    .map_err(ParsingError::into_report)?;

    let project_config_paths = workspace_file
        .project_paths
        .iter()
        .flat_map(|project_path| find_project_files(workspace_path.parent().unwrap(), project_path))
        .collect::<Vec<_>>();

    let project_files = project_config_paths
        .iter()
        .map(|f| {
            let path = f
                .canonicalize()
                .expect("project path to be canonicalizable");

            parse_project_file(
                &path,
                &std::fs::read_to_string(f).expect("couldn't read project file"),
            )
            .map_err(ParsingError::into_report)
        })
        .collect::<Result<Vec<_>, _>>()?;

    Ok((workspace_file, project_files))
}

#[derive(Debug, miette::Diagnostic, thiserror::Error)]
#[error("Couldn't find a workspace.kdl")]
#[diagnostic(help("nabs requires a workspace.kdl in the current directory or a parent directory"))]
struct MissingWorkspaceFile;

fn find_project_files(root: &Path, project_path: &str) -> Vec<PathBuf> {
    let root = root.as_os_str().to_str().unwrap();
    let project_path = match project_path.starts_with('/') {
        true => format!("{root}{project_path}"),
        false => format!("{root}/{project_path}"),
    };
    let glob = match project_path.ends_with('/') {
        true => format!("{project_path}project.kdl"),
        false => format!("{project_path}/project.kdl"),
    };
    // TODO: At some point want to update this to use ignore.  not now though
    glob::glob(&glob)
        .expect("Project path patterns broke innit")
        .map(|r| r.expect("something wrong with the result"))
        .collect()
}

fn find_workspace_file() -> Option<PathBuf> {
    let mut current_path = std::env::current_dir().ok()?;

    while current_path.parent().is_some() {
        current_path.push("workspace.kdl");
        if current_path.exists() {
            return Some(
                current_path
                    .canonicalize()
                    .expect("to be able to canonicalize root path"),
            );
        }
        current_path.pop();
        current_path.pop();
    }

    None
}
