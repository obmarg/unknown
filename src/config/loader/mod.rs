use std::path::{self, Path, PathBuf};

use miette::Context;

use super::{
    parse_project_file, parse_task_file, parse_workspace_file, ParsingError, ProjectFile,
    WorkspaceFile,
};

#[cfg(test)]
mod tests;

pub fn load_config_from_path(
    current_path: PathBuf,
) -> Result<(WorkspaceFile, Vec<ProjectFile>), miette::Report> {
    let current_path = current_path
        .canonicalize()
        .expect("to be able to canonicalise current path");

    let workspace_path = find_workspace_file(current_path).ok_or(MissingWorkspaceFile)?;
    let workspace_file = parse_workspace_file(
        &workspace_path,
        &std::fs::read_to_string(&workspace_path).expect("couldn't read workspace file"),
    )?;

    let workspace_root = workspace_path.parent().unwrap();

    let project_config_paths = workspace_file
        .config
        .project_paths
        .iter()
        .flat_map(|project_path| find_project_files(workspace_root, project_path))
        .collect::<Vec<_>>();

    let project_files = project_config_paths
        .iter()
        .map(|path| {
            let canonical_path = path
                .canonicalize()
                .expect("project path to be canonicalizable");

            let Ok(relative_path) = canonical_path.strip_prefix(workspace_root) else {
                return Err(ProjectImportError::FileNotInWorkspace { file_path: canonical_path, workspace_root: workspace_root.to_owned() })
            };

            let project_file = parse_project_file(
                relative_path,
                &std::fs::read_to_string(&canonical_path).expect("couldn't read project file"),
            ).map_err(|e| ProjectImportError::ParsingError(e, relative_path.to_owned()))?;

            Ok(import_tasks(project_file, workspace_root)?)
        })
        .collect::<Result<Vec<_>, _>>().map_err(ProjectImportError::into_report)?;

    Ok((workspace_file, project_files))
}

#[derive(Debug, miette::Diagnostic, thiserror::Error)]
pub enum ProjectImportError {
    #[error("Couldn't load file {1}: {0}")]
    IoError(std::io::Error, PathBuf),
    #[error("Couldn't parse project file {1}")]
    ParsingError(super::ParsingError, PathBuf),
    #[error("File not contained in workspace")]
    FileNotInWorkspace {
        file_path: PathBuf,
        workspace_root: PathBuf,
    },
    #[error("Error importing a task")]
    #[diagnostic()]
    ErrorImportingTasks(#[from] TaskImportError),
}

impl ProjectImportError {
    pub fn into_report(self) -> miette::Report {
        match self {
            ProjectImportError::ParsingError(inner, _) => miette::Report::new(inner),
            ProjectImportError::ErrorImportingTasks(TaskImportError::ParsingError(inner)) => {
                miette::Report::new(inner)
            }
            other => miette::Report::new(other),
        }
    }
}

#[derive(Debug, miette::Diagnostic, thiserror::Error)]
#[error("Couldn't find a workspace.kdl")]
#[diagnostic(help("nabs requires a workspace.kdl in the current directory or a parent directory"))]
struct MissingWorkspaceFile;

fn find_project_files(root: &Path, project_path: &str) -> Vec<PathBuf> {
    let root = root.as_os_str().to_str().unwrap();
    let project_path = match project_path.starts_with('/') {
        true => format!("{root}{project_path}"),
        false => format!("{root}/{project_path}"),
    };
    let glob = match project_path.ends_with('/') {
        true => format!("{project_path}project.kdl"),
        false => format!("{project_path}/project.kdl"),
    };
    // TODO: At some point want to update this to use ignore.  not now though

    // TODO: probably also need to make sure none of these are canonically out of our workspace...
    glob::glob(&glob)
        .expect("Project path patterns broke innit")
        .map(|r| r.expect("something wrong with the result"))
        .collect()
}

fn find_workspace_file(mut current_path: PathBuf) -> Option<PathBuf> {
    while current_path.parent().is_some() {
        current_path.push("workspace.kdl");
        if current_path.exists() {
            return Some(
                current_path
                    .canonicalize()
                    .expect("to be able to canonicalize root path"),
            );
        }
        current_path.pop();
        current_path.pop();
    }

    None
}

// TODO: ideally want this to reference the place where we failed to load...
#[derive(Debug, miette::Diagnostic, thiserror::Error)]
#[diagnostic(help("There was a problem loading a referenced task."))]
pub enum TaskImportError {
    #[error("Couldn't load {0}: {1}")]
    IoError(PathBuf, std::io::Error),
    #[error("Couldn't parse task file")]
    ParsingError(super::ParsingError),
    #[error("File not contained in workspace")]
    FileNotInWorkspace {
        file_path: PathBuf,
        workspace_root: PathBuf,
    },
}

fn import_tasks(
    mut project_file: ProjectFile,
    workspace_root: &Path,
) -> Result<ProjectFile, TaskImportError> {
    let mut imports = std::mem::take(&mut project_file.config.tasks.imports);
    while let Some(import) = imports.pop() {
        let path = match import.strip_prefix('/') {
            Some(relative_path) => workspace_root.join(relative_path),
            None => workspace_root.join(project_file.project_root.join(import)),
        };

        let Ok(relative_path) = path.strip_prefix(workspace_root) else {
            return Err(TaskImportError::FileNotInWorkspace { file_path: path, workspace_root: workspace_root.to_owned() });
        };

        // TODO: for safety, need to make sure this is still a subpath of workspace_root.
        // possibly also need to canonicalise it...?
        let path = path
            .canonicalize()
            .map_err(|e| TaskImportError::IoError(relative_path.to_owned(), e))?;

        if !path.starts_with(workspace_root) {
            return Err(TaskImportError::FileNotInWorkspace {
                file_path: path,
                workspace_root: workspace_root.to_owned(),
            });
        }

        let parsed = parse_task_file(
            &path,
            &std::fs::read_to_string(&path)
                .map_err(|e| TaskImportError::IoError(relative_path.to_owned(), e))?,
        )
        .map_err(TaskImportError::ParsingError)?;

        imports.extend(parsed.config.imports);

        project_file.config.tasks.tasks.extend(parsed.config.tasks);
    }

    Ok(project_file)
}
