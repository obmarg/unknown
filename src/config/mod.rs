mod project;
mod tasks;
mod workspace;

pub use project::ProjectNode;
pub use workspace::WorkspaceFile;

#[cfg(test)]
mod tests;

// TODO: flesh this out
struct ParsingError(knuffel::Error);

fn project_from_str(s: &str) -> Result<Vec<ProjectNode>, ParsingError> {
    // TODO: parse a project from a string.
    knuffel::parse("file", s).map_err(ParsingError)
}

fn workspace_from_str(s: &str) -> Result<WorkspaceFile, ParsingError> {
    // TODO: parse a project from a string.
    knuffel::parse("file", s).map_err(ParsingError)
}
