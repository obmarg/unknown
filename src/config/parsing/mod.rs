use camino::{Utf8Path, Utf8PathBuf};

mod project;
mod tasks;
mod validation;
mod workspace;

pub(super) use self::{project::ProjectDefinition, tasks::*, workspace::WorkspaceDefinition};

pub use validation::Validator;

use super::ConfigSource;

#[derive(Debug, thiserror::Error)]
#[error("Error parsing {1}")]
pub struct ParsingError(pub(super) knuffel::Error, pub(super) Utf8PathBuf);

impl miette::Diagnostic for ParsingError {
    fn code<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
        self.0.code()
    }

    fn severity(&self) -> Option<miette::Severity> {
        self.0.severity()
    }

    fn help<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
        self.0.help()
    }

    fn url<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
        self.0.url()
    }

    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
        self.0.source_code()
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
        self.0.labels()
    }

    fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn miette::Diagnostic> + 'a>> {
        self.0.related()
    }

    fn diagnostic_source(&self) -> Option<&dyn miette::Diagnostic> {
        self.0.diagnostic_source()
    }
}

pub fn parse_project_file(config: &ConfigSource) -> Result<ProjectDefinition, ParsingError> {
    let filename = config.filename();
    let config = knuffel::parse::<ProjectDefinition>(filename, config.contents())
        .map_err(|e| ParsingError(e, filename.to_owned().into()))?;

    // TODO: At some point want to validate the data in the file.
    // e.g. names can't have commas or slashes in them etc.

    Ok(config)
}

pub fn parse_workspace_file(
    path: &Utf8Path,
    contents: &str,
) -> Result<WorkspaceDefinition, ParsingError> {
    let config = knuffel::parse(
        path.file_name()
            .expect("workspace file path to have a filename"),
        contents,
    )
    .map_err(|e| {
        let filename = Utf8PathBuf::from(path.file_name().unwrap());
        ParsingError(e, filename)
    })?;

    // TODO: At some point want to validate the data in the file.
    // e.g. names can't have commas or slashes in them etc.

    Ok(config)
}

pub fn parse_task_file(path: &Utf8Path, contents: &str) -> Result<TaskBlock, ParsingError> {
    let config = knuffel::parse(
        path.file_name()
            .expect("workspace file path to have a filename"),
        contents,
    )
    .map_err(|e| ParsingError(e, path.to_owned()))?;

    // TODO: At some point want to validate the data in the file.
    // e.g. names can't have commas or slashes in them etc.

    Ok(config)
}
