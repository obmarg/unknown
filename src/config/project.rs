use super::tasks;

#[derive(knuffel::Decode, Debug)]
pub struct ProjectFile {
    #[knuffel(child, unwrap(argument))]
    pub project: String,

    #[knuffel(child)]
    pub dependencies: DependencyBlock,

    #[knuffel(child)]
    pub tasks: tasks::TaskBlock,
}

pub enum ProjectNode {
    Project(Project),
    Dependencies(DependencyBlock),
    Tasks(tasks::TaskBlock),
}

#[derive(knuffel::Decode, Debug)]
pub struct Project {
    #[knuffel(argument)]
    name: String,
}

#[derive(knuffel::Decode, Debug)]
pub struct DependencyBlock {
    #[knuffel(children(name = "project"))]
    projects: Vec<ProjectDependency>,

    #[knuffel(children(name = "path"))]
    paths: Vec<PathDependency>,

    #[knuffel(children(name = "import"))]
    imports: Vec<DependencyImport>,
}

#[derive(knuffel::Decode, Debug)]
pub struct ProjectDependency {
    #[knuffel(argument)]
    name: String,
}

#[derive(knuffel::Decode, Debug)]
pub struct PathDependency {
    #[knuffel(argument)]
    path: String,
}

#[derive(knuffel::Decode, Debug)]
pub struct DependencyImport {
    #[knuffel(argument)]
    path: String,
}
