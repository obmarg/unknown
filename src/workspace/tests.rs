use insta::{dynamic_redaction, Settings};
use similar_asserts::assert_eq;

use crate::config::load_config_from_path;

use super::*;

#[test]
fn snapshot_sample_monorepo() {
    let config = load_config_from_path("src/workspace/test-data/".into()).unwrap();

    let mut workspace = Workspace::new(config.workspace_file);
    workspace.add_projects(config.project_files).unwrap();

    let mut settings = Settings::clone_current();
    settings.add_redaction(
        ".**.root_path",
        dynamic_redaction(|value, _| {
            // assert_eq!(path.to_string(), ".info.root_path");
            assert_eq!(
                value.as_str().unwrap().contains("src/workspace/test-data/"),
                true
            );
            "src/workspace/test-data/"
        }),
    );
    settings.set_sort_maps(true);
    settings.sort_selector(".**.project_indices");
    settings.sort_selector(".**.task_indices");
    // settings.sort_selector(".**.nodes");
    settings.sort_selector(".**.edges");
    settings.sort_selector(".**.task_map");
    settings.sort_selector(".**.project_map");

    settings.bind(
        || insta::assert_json_snapshot!(workspace, {".**.nodes" => insta::sorted_redaction(), ".**.edges" => insta::sorted_redaction()}),
    )
}

#[test]
fn test_task_ref_direct_dependencies() {
    let workspace = a_workspace();

    let build_lib_ref = workspace
        .project_at_path("projects/a-lib")
        .unwrap()
        .lookup_task("build", &workspace)
        .unwrap()
        .task_ref();

    let build_project_ref = workspace
        .project_at_path("projects/a-service")
        .unwrap()
        .lookup_task("build", &workspace)
        .unwrap()
        .task_ref();

    assert_eq!(
        build_project_ref.direct_dependencies(&workspace),
        maplit::hashset! { build_lib_ref }
    )
}

fn a_workspace() -> Workspace {
    let config = load_config_from_path("src/workspace/test-data/".into()).unwrap();

    let mut workspace = Workspace::new(config.workspace_file);
    workspace.add_projects(config.project_files).unwrap();
    workspace
}
