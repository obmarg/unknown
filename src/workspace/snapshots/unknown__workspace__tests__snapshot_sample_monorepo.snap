---
source: src/workspace/tests.rs
expression: workspace
---
{
  "info": {
    "name": "my-workspace",
    "project_paths": [
      "**"
    ],
    "root_path": "src/workspace/test-data/"
  },
  "graph_": {
    "graph": {
      "nodes": [
        "WorkspaceRoot",
        {
          "Project": "projects/a-lib"
        },
        {
          "Project": "projects/a-service"
        },
        {
          "Task": [
            "projects/a-lib",
            "build"
          ]
        },
        {
          "Task": [
            "projects/a-lib",
            "hello"
          ]
        },
        {
          "Task": [
            "projects/a-service",
            "build"
          ]
        },
        {
          "Task": [
            "projects/a-service",
            "bye"
          ]
        }
      ],
      "node_holes": [],
      "edge_property": "directed",
      "edges": [
        [
          0,
          1,
          "HasProject"
        ],
        [
          0,
          2,
          "HasProject"
        ],
        [
          1,
          2,
          "ProjectDependsOn"
        ],
        [
          1,
          3,
          "HasTask"
        ],
        [
          1,
          4,
          "HasTask"
        ],
        [
          2,
          1,
          "ProjectDependedOnBy"
        ],
        [
          2,
          5,
          "HasTask"
        ],
        [
          2,
          6,
          "HasTask"
        ],
        [
          3,
          5,
          "TaskDependsOn"
        ],
        [
          5,
          3,
          "TaskDependedOnBy"
        ]
      ]
    },
    "project_indices": [
      [
        "projects/a-lib",
        2
      ],
      [
        "projects/a-service",
        1
      ]
    ],
    "task_indices": [
      [
        [
          "projects/a-lib",
          "build"
        ],
        5
      ],
      [
        [
          "projects/a-lib",
          "hello"
        ],
        6
      ],
      [
        [
          "projects/a-service",
          "build"
        ],
        3
      ],
      [
        [
          "projects/a-service",
          "bye"
        ],
        4
      ]
    ],
    "root_index": 0
  },
  "project_map": [
    [
      "projects/a-lib",
      {
        "name": "a-lib",
        "dependencies": [],
        "root": "projects/a-lib"
      }
    ],
    [
      "projects/a-service",
      {
        "name": "a-service",
        "dependencies": [
          "projects/a-lib"
        ],
        "root": "projects/a-service"
      }
    ]
  ],
  "task_map": [
    [
      [
        "projects/a-lib",
        "build"
      ],
      {
        "project": "projects/a-lib",
        "project_name": "a-lib",
        "name": "build",
        "commands": [
          "echo \"build-a-lib\""
        ],
        "inputs": {
          "paths": [],
          "env_vars": [],
          "commands": []
        }
      }
    ],
    [
      [
        "projects/a-lib",
        "hello"
      ],
      {
        "project": "projects/a-lib",
        "project_name": "a-lib",
        "name": "hello",
        "commands": [
          "echo \"hello\""
        ],
        "inputs": {
          "paths": [],
          "env_vars": [],
          "commands": []
        }
      }
    ],
    [
      [
        "projects/a-service",
        "build"
      ],
      {
        "project": "projects/a-service",
        "project_name": "a-service",
        "name": "build",
        "commands": [
          "echo \"build-a-service\""
        ],
        "inputs": {
          "paths": [],
          "env_vars": [],
          "commands": []
        }
      }
    ],
    [
      [
        "projects/a-service",
        "bye"
      ],
      {
        "project": "projects/a-service",
        "project_name": "a-service",
        "name": "bye",
        "commands": [
          "echo \"bye\""
        ],
        "inputs": {
          "paths": [],
          "env_vars": [],
          "commands": []
        }
      }
    ]
  ],
  "task_requirements": [
    [
      [
        "projects/a-service",
        "build"
      ],
      [
        [
          "projects/a-lib",
          "build"
        ]
      ]
    ],
    [
      [
        "projects/a-service",
        "bye"
      ],
      []
    ],
    [
      [
        "projects/a-lib",
        "build"
      ],
      []
    ],
    [
      [
        "projects/a-lib",
        "hello"
      ],
      []
    ]
  ]
}
